Python/tier2_redundancy_eliminator_bytecodes.c

```diff
@@ -81,12 +81,62 @@ dummy_func(void) {
 
 
     op(_BINARY_OP_ADD_INT, (left, right -- res)) {
-        // TODO constant propagation
-        (void)left;
-        (void)right;
-        res = sym_new_known_type(ctx, &PyLong_Type);
-        if (res == NULL) {
-            goto out_of_space;
+        if (is_const(left) && is_const(right)) {
+            assert(PyLong_CheckExact(get_const(left)));
+            assert(PyLong_CheckExact(get_const(right)));
+            PyObject *temp = _PyLong_Add((PyLongObject *)get_const(left),
+                                         (PyLongObject *)get_const(right));
+            if (temp == NULL) {
+                goto error;
+            }
+            res = sym_new_const(ctx, temp);
+            // TODO replace opcode with constant propagated one and add tests!
```

Comment:
Does this mean it's premature to review? In that case could you turn it into Draft?

Reply:
No it's ready. I want to do the constant emission in a separate PR, because of the `_SHRINK_STACK` issue that you pointed out previously.

The current infrastructure uses the same trace buffer as a write buffer, so it's possible for `_SHRINK_STACK` to overshoot the buffer and corrupt it. I want to solve that in another PR.